### PR TITLE
expression: Fix incorrect behavior of greatest/least when mixed temporal type with others

### LINKED ISSUE #30264: GREATEST function with plus operations to time/date type gives wrong result.
### 1. Minimal reproduce step 

```sql
SELECT GREATEST(time '20:00:00', 120000) + 0.00;
SELECT GREATEST(date '2005-05-05', 20010101, 20040404, 20030303) + 0;
SELECT GREATEST(date '1995-05-05', 19910101, 20050505, 19930303) + 0.00;
```

### 2. What did you expect to see? 
```sql
mysql> SELECT GREATEST(time '20:00:00', 120000) + 0.00;
+------------------------------------------+
| GREATEST(time '20:00:00', 120000) + 0.00 |
+------------------------------------------+
|                                200000.00 |
+------------------------------------------+
1 row in set (0.00 sec)

mysql> SELECT GREATEST(date '2005-05-05', 20010101, 20040404, 20030303) + 0;
+---------------------------------------------------------------+
| GREATEST(date '2005-05-05', 20010101, 20040404, 20030303) + 0 |
+---------------------------------------------------------------+
|                                                      20050505 |
+---------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> SELECT GREATEST(date '1995-05-05', 19910101, 20050505, 19930303) + 0.00;
+------------------------------------------------------------------+
| GREATEST(date '1995-05-05', 19910101, 20050505, 19930303) + 0.00 |
+------------------------------------------------------------------+
|                                                      20050505.00 |
+------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### 3. What did you see instead 
```sql

TiDB> SELECT GREATEST(time '20:00:00', 120000) + 0.00;
+------------------------------------------+
| GREATEST(time '20:00:00', 120000) + 0.00 |
+------------------------------------------+
|                                       20 |
+------------------------------------------+
1 row in set, 1 warning (0.00 sec)

TiDB> SELECT GREATEST(date '2005-05-05', 20010101, 20040404, 20030303) + 0;
+---------------------------------------------------------------+
| GREATEST(date '2005-05-05', 20010101, 20040404, 20030303) + 0 |
+---------------------------------------------------------------+
|                                                          2005 |
+---------------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

TiDB> SELECT GREATEST(date '1995-05-05', 19910101, 20050505, 19930303) + 0.00;
+------------------------------------------------------------------+
| GREATEST(date '1995-05-05', 19910101, 20050505, 19930303) + 0.00 |
+------------------------------------------------------------------+
|                                                             2005 |
+------------------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)
```

### 4. What is your TiDB version? (Required)

```sql
tidb_version(): Release Version: v5.2.2
Edition: Community
Git Commit Hash: da1c21fd45a4ea5900ac16d2f4a248143f378d18
Git Branch: heads/refs/tags/v5.2.2
UTC Build Time: 2021-10-20 06:03:29
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by yibin87**
### What problem does this PR solve?

Issue Number: close #30264 #30261

Problem Summary:

### What is changed and how it works?
Change Greatest/Least builtin function's behavior according to MySQL's. 
1. Handle different type mixtures: Date/Time/DateTime/String/Int.
2. Separate Date Comparison from Time Comparison, since they return different types.
### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by yibin87** [comment]
\run-all-tests

File: expression/builtin_compare.go [hunk 1]
```diff
@@ -963,6 +1034,17 @@ func (b *builtinLeastTimeSig) evalTime(row chunk.Row) (res types.Time, isNull bo
 			res = v
 		}
 	}
+	// Convert ETType Time value to MySQL actual type, distinguish date and datetime
+	sc := b.ctx.GetSessionVars().StmtCtx
+	var resTimeTp byte
+	if b.cmpAsDate {
+		resTimeTp = mysql.TypeDate
```
Comment (guo-shaoge) [inline expression/builtin_compare.go:1041]: Why not use `b.tp` to as resTimeTp?

Comment (yibin87) [reply expression/builtin_compare.go:1041]: Currently, b.tp is generated from EType, which ignores the differences between date and datetime.

**Review by guo-shaoge — APPROVED** [review]

File: expression/builtin_compare.go [hunk 2]
```diff
@@ -417,7 +417,10 @@ func ResolveType4Between(args [3]Expression) types.EvalType {
 }
 
 // resolveType4Extremum gets compare type for GREATEST and LEAST and BETWEEN (mainly for datetime).
-func resolveType4Extremum(args []Expression) (_ types.EvalType, cmpStringAsDatetime bool) {
+// cmpStringMode: 0 for cmp string directly; 1 for cmp string as date 'yyyy-mm-dd'; 2 for cmp string
+// as datetime 'yyyy-mm-dd hh:mm:ss'
+// fieldTimeType: the return type, 0 for non-temporary time; 1 for date type; 2 for datetime type
+func resolveType4Extremum(args []Expression) (_ types.EvalType, fieldTimeType byte, cmpStringMode int) {
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go:423]:
```suggestion
type CmpStringMode uint8

const (
	CmpStringDirectly CmpStringMode = iota
	CmpStringAsDate // 'yyyy-mm-dd'
	CmpStringAsDatetime // 'yyyy-mm-dd hh:mm:ss' 
)

func resolveType4Extremum(args []Expression) (_ types.EvalType, fieldTimeType byte, cmpStringMode CmpStringMode) {
```

Comment (XuHuaiyu) [reply expression/builtin_compare.go:423]: the `fieldTimeType` can also be extracted as an individual type definition

File: expression/builtin_compare.go [hunk 3]
```diff
@@ -945,11 +1014,13 @@ func (b *builtinLeastCmpStringAsTimeSig) evalString(row chunk.Row) (strRes strin
 
 type builtinLeastTimeSig struct {
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go:1015]: The `vecEvalxxx` of the sig also needs to be fixed.

File: expression/builtin_compare.go [hunk 4]
```diff
@@ -18,16 +18,18 @@ import (
 	"math"
 	"strings"
 
+	"github.com/pingcap/tipb/go-tipb"
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go:21]: move this line to line 33

**Review by XuHuaiyu — APPROVED** [review]
LGTM

**Comment by XuHuaiyu** [comment]
related to https://github.com/pingcap/tidb/pull/29408

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**